server/src/main/java/org/apache/iotdb/db/query/dataset/NewEngineDataSetWithoutValueFilter.java

```diff
@@ -327,6 +328,167 @@ public TSQueryDataSet fillBuffer(int fetchSize, WatermarkEncoder encoder) throws
+            ReadWriteIOUtils.write(time, timeBAOSList[seriesIndex]);
+            currentBitmapList[seriesIndex] = (currentBitmapList[seriesIndex] << 1) | FLAG;
+            TSDataType type = cachedBatchDataArray[seriesIndex].getDataType();
+            switch (type) {
+              case INT32:
+                int intValue = cachedBatchDataArray[seriesIndex].getInt();
+                if (encoder != null && encoder.needEncode(time)) {
+                  intValue = encoder.encodeInt(intValue, time);
+                }
+                ReadWriteIOUtils.write(intValue, valueBAOSList[seriesIndex]);
+                break;
+              case INT64:
+                long longValue = cachedBatchDataArray[seriesIndex].getLong();
+                if (encoder != null && encoder.needEncode(time)) {
+                  longValue = encoder.encodeLong(longValue, time);
+                }
+                ReadWriteIOUtils.write(longValue, valueBAOSList[seriesIndex]);
+                break;
+              case FLOAT:
+                float floatValue = cachedBatchDataArray[seriesIndex].getFloat();
+                if (encoder != null && encoder.needEncode(time)) {
+                  floatValue = encoder.encodeFloat(floatValue, time);
+                }
+                ReadWriteIOUtils.write(floatValue, valueBAOSList[seriesIndex]);
+                break;
+              case DOUBLE:
+                double doubleValue = cachedBatchDataArray[seriesIndex].getDouble();
+                if (encoder != null && encoder.needEncode(time)) {
+                  doubleValue = encoder.encodeDouble(doubleValue, time);
+                }
+                ReadWriteIOUtils.write(doubleValue, valueBAOSList[seriesIndex]);
+                break;
+              case BOOLEAN:
+                ReadWriteIOUtils.write(cachedBatchDataArray[seriesIndex].getBoolean(),
+                        valueBAOSList[seriesIndex]);
+                break;
+              case TEXT:
+                ReadWriteIOUtils
+                        .write(cachedBatchDataArray[seriesIndex].getBinary(),
+                                valueBAOSList[seriesIndex]);
+                break;
+              default:
+                throw new UnSupportedDataTypeException(
+                        String.format("Data type %s is not supported.", type));
+            }
+          }
+
+          // move next
+          cachedBatchDataArray[seriesIndex].next();
+
+          // get next batch if current batch is empty
+          if (!cachedBatchDataArray[seriesIndex].hasCurrent()) {
+            // still have remaining batch data in queue
+            if (!noMoreDataInQueueArray[seriesIndex]) {
+              fillCache(seriesIndex);
+            }
+          }
+
+          // try to put the next timestamp into the heap
+          if (cachedBatchDataArray[seriesIndex].hasCurrent()) {
+            long time = cachedBatchDataArray[seriesIndex].currentTime();
+            timeHeap.add(time);
```

Comment:
I wonder if it is necessary to use a heap for the non-align query.

Reply:
Actually, i think it's a bug to use timeheap here

Reply:
Yes, the timeheap is not necessary. Removed.